sentry-actix/src/lib.rs

```diff
@@ -68,221 +71,267 @@
 #![warn(missing_docs)]
 #![allow(clippy::needless_doctest_main)]
 #![allow(deprecated)]
+#![allow(clippy::type_complexity)]
 
 use std::borrow::Cow;
-use std::cell::RefCell;
-use std::sync::{Arc, Mutex};
+use std::pin::Pin;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
+use actix_web::{Error, HttpMessage, HttpRequest};
+use futures_util::future::{ok, Future, Ready};
+use futures_util::FutureExt;
 
-use actix_web::middleware::{Finished, Middleware, Response, Started};
-use actix_web::{Error, HttpMessage, HttpRequest, HttpResponse};
-use failure::Fail;
-use sentry::protocol::{ClientSdkPackage, Event, Level};
+use sentry::protocol::{ClientSdkPackage, Event};
 use sentry::types::Uuid;
-use sentry::{Hub, ScopeGuard};
-use sentry_failure::exception_from_single_fail;
+use sentry::Hub;
 
 /// A helper construct that can be used to reconfigure and build the middleware.
-pub struct SentryMiddlewareBuilder {
-    middleware: SentryMiddleware,
+pub struct SentryBuilder {
+    middleware: Sentry,
 }
 
-impl SentryMiddlewareBuilder {
+impl SentryBuilder {
     /// Finishes the building and returns a middleware
-    pub fn finish(self) -> SentryMiddleware {
+    pub fn finish(self) -> Sentry {
         self.middleware
     }
 
     /// Reconfigures the middleware so that it uses a specific hub instead of the default one.
     pub fn with_hub(mut self, hub: Arc<Hub>) -> Self {
-        self.middleware.hub = Some(hub);
+        let inner = Rc::get_mut(&mut self.middleware.0).expect("Multiple copies exist");
+        inner.hub = Some(hub);
         self
     }
 
     /// Reconfigures the middleware so that it uses a specific hub instead of the default one.
     pub fn with_default_hub(mut self) -> Self {
-        self.middleware.hub = None;
+        let inner = Rc::get_mut(&mut self.middleware.0).expect("Multiple copies exist");
+        inner.hub = None;
         self
     }
 
     /// If configured the sentry id is attached to a X-Sentry-Event header.
     pub fn emit_header(mut self, val: bool) -> Self {
-        self.middleware.emit_header = val;
+        let inner = Rc::get_mut(&mut self.middleware.0).expect("Multiple copies exist");
+        inner.emit_header = val;
         self
     }
 
     /// Enables or disables error reporting.
     ///
     /// The default is to report all errors.
     pub fn capture_server_errors(mut self, val: bool) -> Self {
-        self.middleware.capture_server_errors = val;
+        let inner = Rc::get_mut(&mut self.middleware.0).expect("Multiple copies exist");
+        inner.capture_server_errors = val;
         self
     }
 }
 
-/// Reports certain failures to sentry.
-pub struct SentryMiddleware {
+struct Inner {
     hub: Option<Arc<Hub>>,
     emit_header: bool,
     capture_server_errors: bool,
 }
 
+/// Reports certain failures to Sentry.
+pub struct Sentry(Rc<Inner>);
```

Comment:
Its just super weird seeing `Rc` in async code… I think its fine because of `Pin`?
Otherwise, I think its fine to just derive `Clone` here instead of using the Rc.